### PR TITLE
Add ability to check backend readiness

### PR DESCRIPTION
**Pull request by Anton-Kalpakchiev**
As described in #371, all services should implement a "readiness" endpoint, which checks whether the service is ready to serve traffic, i.e. whether it can reach all of its dependencies. Thus, the origin and build index's readiness endpoints should check if they can reach remote backends. This diff adds a function to the backend manager that allows exactly that by iterating through the storage backends and running a Stat call on all of them.

I suggest it be configurable which backends need to be checked by the readiness endpoint (through the `must_ready` flag in the .yaml config), as some are more important for Kraken than others.

### CODE REVIEW

File: lib/backend/manager.go [hunk 1]
```diff
@@ -139,3 +144,18 @@ func (m *Manager) GetClient(namespace string) (Client, error) {
...
+
+// IsReady returns whether the backends are ready (reachable).
+// A backend must be explicitly configured as required for readiness to be checked.
+func (m *Manager) IsReady() (bool, error) {
+	for _, b := range m.backends {
+		if !b.mustReady {
+			continue
+		}
+		_, err := b.client.Stat(isReadyNamespace, isReadyName)
+		if err != nil && err != backenderrors.ErrBlobNotFound {
```
Comment (xinlongz1) [inline lib/backend/manager.go:154]: So is backenderrors.ErrBlobNotFound return for all backends when the blob doesn't exist?

File: lib/backend/manager.go [hunk 2]
```diff
@@ -139,3 +144,18 @@ func (m *Manager) GetClient(namespace string) (Client, error) {
...
 }
+
+// IsReady returns whether the backends are ready (reachable).
+// A backend must be explicitly configured as required for readiness to be checked.
+func (m *Manager) IsReady() (bool, error) {
+	for _, b := range m.backends {
+		if !b.mustReady {
+			continue
+		}
+		_, err := b.client.Stat(isReadyNamespace, isReadyName)
```
Comment (xinlongz1) [inline lib/backend/manager.go]: I think we need to clarify the goal of the readiness endpoint: to check backends reachable or backend function correctly?

If the goal is to verify backends are reachable, how about just pinging or HTTP check the backend endpoints?

Comment (Anton-Kalpakchiev) [reply lib/backend/manager.go]: The stronger the check better. I've reworked the check to instead download an empty file. Unfortunately, this requires the Kraken user to manually upload an empty file before being able to use the endpoint.

File: lib/backend/manager.go (hunk 1 above)
Comment (Anton-Kalpakchiev) [reply lib/backend/manager.go:154]: Yes. It's required for the interface. Either way, I've changed the code not to use it anymore.

File: lib/backend/constants.go [hunk 3]
```diff
@@ -24,4 +24,8 @@ const (
 	DefaultBufferGuard datasize.ByteSize = 10 * datasize.MB
 	DefaultConcurrency int               = 10
 	DefaultListMaxKeys int               = 250
+
+	// Name and namespace of an empty blob used to check readiness.
+	ReadinessNamespace string = "readiness/blob"
+	ReadinessName      string = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
```
Comment (gkeesh7) [inline lib/backend/constants.go]: nit :- `ReadineessCheckBlobHash` seems more descriptive ?

Comment (Anton-Kalpakchiev) [reply lib/backend/constants.go]: I agree. Changed it.

File: lib/backend/manager.go [hunk 4]
```diff
@@ -139,3 +142,18 @@ func (m *Manager) GetClient(namespace string) (Client, error) {
...
 }
+
+// CheckReadiness returns whether the backends are ready (by trying to download a file).
+// A backend must be explicitly configured as required for readiness to be checked.
+func (m *Manager) CheckReadiness() error {
+	for _, backend := range m.backends {
+		if !backend.mustReady {
+			continue
+		}
+		err := backend.client.Download(ReadinessCheckBlobNamespace, ReadinessCheckBlobName, bytes.NewBuffer([]byte{}))
```
Comment (gkeesh7) [inline lib/backend/manager.go]: Let's not use fixed blob hashes to check for backend readiness

Comment (Anton-Kalpakchiev) [reply lib/backend/manager.go]: Reworked.

**Review by gkeesh7 — APPROVED** [review]

**Pull request merged**